File: python/src/day06.py
#!/usr/bin/env python

from itertools import chain
from pathlib import Path
from typing import List

INPUT_FILE = Path.cwd().parent / "inputs" / "day06.txt"

AnswerGroup = List[str]


def transform_input(input: str) -> List[AnswerGroup]:
    return [x.split("\n") for x in input.strip().split("\n\n")]


def part1(groups: List[AnswerGroup]) -> int:
    return sum(len(set(chain(*x))) for x in groups)


def part2(groups: List[AnswerGroup]) -> int:
    return sum(len(set.intersection(*[set(y) for y in x])) for x in groups)


if __name__ == "__main__":
    with INPUT_FILE.open("r") as f:
        input = transform_input(f.read())

    print(f"Part 1: {part1(input)}")
    print(f"Part 2: {part2(input)}")

File: rust/src/day06.rs
use std::collections::HashSet;

static INPUT: &str = include_str!("../../inputs/day06.txt");

type AnswerGroup<'a> = Vec<&'a str>;

pub fn run() {
    let groups = transform_input(INPUT);

    println!("Part 1: {}", part1(&groups));
    println!("Part 2: {}", part2(&groups));
}

fn transform_input(input: &str) -> Vec<AnswerGroup> {
    input
        .trim_end()
        .split("\n\n")
        .map(|x| x.split('\n').collect())
        .collect()
}

fn part1(groups: &[AnswerGroup]) -> usize {
    groups
        .iter()
        .map(|answers| {
            answers
                .iter()
                .flat_map(|x| x.chars())
                .collect::<HashSet<_>>()
                .len()
        })
        .sum()
}

fn part2(groups: &[AnswerGroup]) -> usize {
    groups
        .iter()
        .map(|answers| {
            answers
                .iter()
                .map(|x| x.chars().collect::<HashSet<_>>())
                .fold_first(|acc, x| acc.intersection(&x).cloned().collect())
                .unwrap()
                .len()
        })
        .sum()
}
